File: 23.2-completable-future/2-exception/js/Exception.js
function delayedFailure(value, delay) {
	return new Promise((resolve, reject) => {
		// after 'delay' milliseconds, reject the promise with 'value'
		setTimeout(() => { reject(value) }, delay)
	})
}

const promise = delayedFailure(1234, 3000)

console.log(":: promise created ::")

/*
// Process exceptions using 'catch'
promise.catch(value => {
	// will run only after the promise is rejected
	console.log(`-- failure data: ${ value } --`)
})
*/

// Process exceptions using 'then' with two callbacks,
// one for success and another for failure
promise.then(
	value => {
		// will run only if the promise becomes fulfilled
		console.log(`++ value produced: ${ value } ++`)
	},
	exception => {
		// will run only if the promise becomes rejected
		console.log(`-- failure data: ${ exception } --`)
	}
)

console.log(":: callback registered ::")

console.log(":: waiting... ::")

File: 23.2-completable-future/2-exception/kt/Exception.kt
import java.util.concurrent.CompletableFuture
import kotlin.concurrent.thread

fun delayedFailure(value: Int, delay: Long) : CompletableFuture<Int> {
	val res = CompletableFuture<Int>()

	thread {
		// after 'delay' milliseconds, complete the future with an exception
		// carrying 'value' as a message
		Thread.sleep(delay)
		res.completeExceptionally(Exception(value.toString()))
	}

	return res
}

fun main() {
	val future = delayedFailure(1234, 3000)

	println(":: promise created ::")

	// Process exceptions using 'whenComplete' and a callback with two parameters,
	// one for a value (on success) and another for an exception (on failure)
	future.whenComplete { value, exception ->
		// will run only after the future is completed (normally or exceptionally)
		if (exception == null) {
			println("++ value produced: ${ value } ++")
		} else {
			println("-- failure data: ${ exception.message } --")
		}
	}

	println(":: callback registered ::")

	println(":: waiting... ::")
}
